### PR TITLE
Add timeouts support to vs/vsr

### PR DESCRIPTION
**Pull request by Dean-Coakley**
### Proposed changes
Support configuring timeouts in upstreams of VirtualServer/VirtualServerRoute

Add support for the following when using to vs/vsrs:
* connect-timeout
* read-timeout 
* send-timeout

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -29,9 +29,9 @@ lint:
 
 test:
 ifeq ($(BUILD_IN_CONTAINER),1)
-	$(DOCKER_RUN) $(GOLANG_CONTAINER) go test ./...
+	$(DOCKER_RUN) $(GOLANG_CONTAINER) go test -race ./...
```
Comment (pleshakov) [inline Makefile]: this change and the one in line 34 have nothing to do with the PR. please remove.

File: docs/virtualserver-and-virtualserverroute.md [hunk 2]
```diff
@@ -189,6 +192,9 @@ max-fails: 1
...
 | `max-fails` | The number of unsuccessful attempts to communicate with an upstream server that should happen in the duration set by the `fail-timeout` to consider the server unavailable. See the [max_fails](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#max_fails) parameter of the server directive. The default is set in the `max-fails` ConfgMap key. | `int` | No |
+`connect-timeout` | The timeout for establishing a connection with an upstream server. See the [proxy_connect_timeout](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_connect_timeout) directive. The default is specified in the `proxy-connect-timeout` ConfigMap key. | `string` | No
+`read-timeout` | The timeout for reading a response from an upstream server. See the [proxy_read_timeout](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout) directive.  The default is specified in the `proxy-read-timeout` ConfigMap key. | `string` | No
+`send-timeout` | The timeout for transmitting a request to an upstream server. See the [proxy_send_timeout](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_send_timeout) directive. The default is 60s. | `string` | No
```
Comment (pleshakov) [inline docs/virtualserver-and-virtualserverroute.md]:
```suggestion
`send-timeout` | The timeout for transmitting a request to an upstream server. See the [proxy_send_timeout](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_send_timeout) directive. The default is `60s`. | `string` | No
```

File: docs/virtualserver-and-virtualserverroute.md [hunk 3]
```diff
@@ -189,6 +192,9 @@ max-fails: 1
...
 | `fail-timeout` | The time during which the specified number of unsuccessful attempts to communicate with an upstream server should happen to consider the server unavailable. See the [fail_timeout](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#fail_timeout) parameter of the server directive. The default is set in the `fail-timeout` ConfigMap key. | `string` | No |
 | `max-fails` | The number of unsuccessful attempts to communicate with an upstream server that should happen in the duration set by the `fail-timeout` to consider the server unavailable. See the [max_fails](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#max_fails) parameter of the server directive. The default is set in the `max-fails` ConfgMap key. | `int` | No |
+`connect-timeout` | The timeout for establishing a connection with an upstream server. See the [proxy_connect_timeout](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_connect_timeout) directive. The default is specified in the `proxy-connect-timeout` ConfigMap key. | `string` | No
```
Comment (pleshakov) [inline docs/virtualserver-and-virtualserverroute.md]: For the added links, use `https://nginx.org` for consistency with the existing links

File: internal/configs/version1/config.go [hunk 4]
```diff
@@ -105,6 +105,7 @@ type Location struct {
 	Upstream             Upstream
 	ProxyConnectTimeout  string
 	ProxyReadTimeout     string
+	ProxySendTimeout     string
```
Comment (pleshakov) [inline internal/configs/version1/config.go]: this is not required. the scope of the PR is for VS/VSR, not Ingress resources.

File: internal/configs/ingress_test.go [hunk 5]
```diff
@@ -161,6 +161,7 @@ func createExpectedConfigForCafeIngressEx() version1.IngressNginxConfig {
 						Upstream:            coffeeUpstream,
 						ProxyConnectTimeout: "60s",
 						ProxyReadTimeout:    "60s",
+						ProxySendTimeout:    "60s",
```
Comment (pleshakov) [inline internal/configs/ingress_test.go]: all the changes in this file are not required. See the comment in the `internal/configs/version1/config.go` file.

File: internal/configs/version2/config.go [hunk 6]
```diff
@@ -11,9 +11,12 @@ type VirtualServerConfig struct {
 
 // Upstream defines an upstream.
 type Upstream struct {
-	Name     string
-	Servers  []UpstreamServer
-	LBMethod string
+	Name                string
+	Servers             []UpstreamServer
+	LBMethod            string
+	ProxyConnectTimeout string
```
Comment (pleshakov) [inline internal/configs/version2/config.go]: lines 17-19 are not necessary

File: pkg/apis/configuration/validation/validation_test.go [hunk 7]
```diff
@@ -3,10 +3,9 @@ package validation
 import (
```
Comment (pleshakov) [inline pkg/apis/configuration/validation/validation_test.go]: no need to change this file

File: internal/configs/virtualserver.go [hunk 8]
```diff
@@ -401,14 +419,14 @@ func generateRulesRouteConfig(route conf_v1alpha1.Route, upstreamNamer *upstream
...
 
-var specialMapParameters = map[string]bool{
-	"default":   true,
-	"hostnames": true,
-	"include":   true,
-	"volatile":  true,
-}
-
 func generateValueForRulesRouteMap(matchedValue string) (value string, isNegative bool) {
+	var specialMapParameters = map[string]bool{
```
Comment (pleshakov) [inline internal/configs/virtualserver.go]: this map is a constant. there is no need to put in the `generateValueForRulesRouteMap` func

File: internal/configs/ingress.go [hunk 9]
```diff
@@ -235,6 +235,7 @@ func createLocation(path string, upstream version1.Upstream, cfg *ConfigParams,
 		Upstream:             upstream,
 		ProxyConnectTimeout:  cfg.ProxyConnectTimeout,
 		ProxyReadTimeout:     cfg.ProxyReadTimeout,
+		ProxySendTimeout:     cfg.ProxySendTimeout,
```
Comment (pleshakov) [inline internal/configs/ingress.go]: this is not required. see the comment in the `internal/configs/version1/config.go` file

File: internal/configs/virtualserver.go (hunk 8 above)
Comment (Rulox) [inline internal/configs/virtualserver.go]: What is the reason to change this to inside the function? I am curious 🤔

File: internal/configs/virtualserver.go [hunk 10]
```diff
@@ -81,9 +81,22 @@ func (namer *variableNamer) GetNameForVariableForRulesRouteMainMap(rulesIndex in
...
+	}
+	return conf_v1alpha1.Upstream{}
+}
+
 func generateVirtualServerConfig(virtualServerEx *VirtualServerEx, tlsPemFileName string, baseCfgParams *ConfigParams, isPlus bool) version2.VirtualServerConfig {
 	ssl := generateSSLConfig(virtualServerEx.VirtualServer.Spec.TLS, tlsPemFileName, baseCfgParams)
 
+	// oUpstreams keeps track of conf_v1alpha1.Upstreams as they are created
+	// necessary to know what Location references which Upstream
+	var oUpstreams []conf_v1alpha1.Upstream
```
Comment (pleshakov) [inline internal/configs/virtualserver.go]: use a map to store upstreams. this way you won't need to have `getUpstream` function.

the key of the map should be the generated upstream name. That name is unique among all upstream. Right now there is a bug in your code: consider the case when you have an upstream with the same name  in VS and VSR.

regarding the variable name, I suggest calling it `crUpstreams`, where cr stands fo custom resources.

File: internal/configs/virtualserver.go (hunk 8 above)
Comment (Dean-Coakley) [reply internal/configs/virtualserver.go]: I simply don't like to have global variables unless there's a good reason for them. This seems only used within this func. I think it's also slightly easier to read.

Would you like me to revert this? - Why?

Comment (pleshakov) [reply internal/configs/virtualserver.go]: it is a variable but it is a constant. we never change it. there is no need to initialize it every time that function is called.

**Pull request merged**